roofit/batchcompute/src/Initialisation.cxx

```diff
@@ -1,75 +1,72 @@
-#include "RooBatchCompute.h"
+#include "rbc.h"
 
 #include "TEnv.h"
 #include "TSystem.h"
+#include "TError.h"
 
-#include <iostream>
 #include <string>
 #include <exception>
 
-
-// First initialisation of the pointer. When implementations of the batch compute library are loaded,
-// they will overwrite the pointer.
-RooBatchCompute::RooBatchComputeInterface* RooBatchCompute::dispatch=nullptr;
+// First initialisation of the pointers. When implementations of the batch compute library
+// are loaded, they will overwrite the pointers.
+RooBatchCompute::RbcInterface *RooBatchCompute::dispatchCPU = nullptr;
+RooBatchCompute::RbcInterface *RooBatchCompute::dispatchCUDA = nullptr;
 
 namespace {
+  
+/// Dynamically load a library and throw exception in case of failure
+void loadWithErrorChecking(const std::string& libName)
+{
+  const auto returnValue = gSystem->Load(libName.c_str());
+  if (returnValue == -1 || returnValue == -2)
+    throw std::runtime_error("RooFit was unable to load its computation library " + libName);
+  else if (returnValue == 1) // Library should not have been loaded before we tried to do it.
+    throw std::logic_error("RooFit computation library " + libName + " was loaded before RooFit initialisation began.");
+}
 
-////////////////////////////////////////////////////////////////////////////////////////////////////////////////
-/// Inspect cpu capabilities, and load the optimal library for RooFit computations.
-void loadComputeLibrary() {
-
-  std::string libName="libRooBatchCompute_GENERIC";
-
+/// Inspect hardware capabilities, and load the optimal library for RooFit computations.
+void loadComputeLibrary()
+{
+  const std::string userChoice = gEnv->GetValue("RooFit.BatchCompute","auto");
```

Comment:
Correct! That's how it works. The CPU architecture is chosen here via `gEnv`, and if CUDA is used or not is decided by the `BatchCompute()` command argument to `RooAbsPdf::fitTo()`.

Reply:
Thanks. And we want to load the CUDA libraries even if it's deselected through `BatchCompute()`?

Reply:
You are right, then you don't need it. Of course we don't know at RooFit library loading if the user is gonna do that, but I'll think of a mechanism to lazy load the CUDA libraries the first time a fit is done in CUDA!

Reply:
Right so in principle this can go into a subsequent PR. I'm just a bit worried that with this PR here, any use of RooFit will currently load the CUDA library if it was found at build time, and IIUC there is no way for the user to prevent that. I don't know whether that can be a problem. Your call! :-)